### ISSUE TITLE
clang-tidy's readability-redundant-member-init has false positives for aggregates

### ISSUE THREAD

**Issue by foonathan** [llvm/llvm-project#119094]
The following code generates a clang-tidy warning for `readability-redundant-member-init`.

```cpp
struct foo {
    std::string a;
    std::string b  = {};
};
```

However, the initializer for `b` is not redundant. Consider:

```cpp
void expect_foo(const foo& f);

expect_foo({"hello"});
```

With the initializer for `b` present, the code is fine. Without the initializer, the code generates a `missing-field-initializer` warning.

**Comment by chrchr-github** [comment]
Seems like this is a problem with `missing-field-initializer` then. That check should probably only warn for types that require initialization.
https://godbolt.org/z/Toscx3rsP

**Comment by chrchr-github** [comment]
Related to/duplicate of https://github.com/llvm/llvm-project/issues/90235

**Comment by foonathan** [comment]
>That check should probably only warn for types that require initialization.

I'd disagree. I rely on that warning to force me to update all call sites when I add a new member.

**Comment by 5chmidti** [comment]
I'm closing this as a duplicate of #90235. 

Here's what I think the resolution of this should be: https://github.com/llvm/llvm-project/issues/90235#issuecomment-2540447364

>I think that the check and the warning are correct in what they diagnose. The check detects two kinds of issues. First, unnecessary initializers in constructors, and unnecessary initializers at the member-declaration. I think it would make sense to allow to configure what the check should detect. For the provided example, you'd disable the checking for redundant field initializations, while keeping the redundant constructor field initialization diagnostics.

**Comment by Denzeli** [comment]
missing-field-initializers does not complain about member-initialized fields of built-in types but warns about fields of class types with user-provided default constructors, although the very meaning of the default constructor is to default initialize class fields. Have to disable the warning because it seems broken the way it is now.
